PxMatrix.cpp

```diff
@@ -0,0 +1,955 @@
+  }
+}
+
+void PxMATRIX::drawPixelRGB565(int16_t x, int16_t y, uint16_t color) {
+  uint8_t r = ((((color >> 11) & 0x1F) * 527) + 23) >> 6;
+  uint8_t g = ((((color >> 5) & 0x3F) * 259) + 33) >> 6;
+  uint8_t b = (((color & 0x1F) * 527) + 23) >> 6;
+#ifdef PxMATRIX_double_buffer
+  fillMatrixBuffer(x, y, r, g, b, !_active_buffer);
+#else
+  fillMatrixBuffer(x, y, r, g, b, false);
+#endif
+}
+
+void PxMATRIX::drawPixelRGB888(int16_t x, int16_t y, uint8_t r, uint8_t g, uint8_t b) {
+#ifdef PxMATRIX_double_buffer
+  fillMatrixBuffer(x, y, r, g, b, !_active_buffer);
+#else
+  fillMatrixBuffer(x, y, r, g, b, false);
+#endif
+}
+
+// the most basic function, get a single pixel
+uint8_t PxMATRIX::getPixel(int8_t x, int8_t y) {
+  return (0);  // PxMATRIX_buffer[x+ (y/8)*LCDWIDTH] >> (y%8)) & 0x1;
+}
+
+void PxMATRIX::begin() { begin(8); }
+
+void PxMATRIX::begin(uint8_t row_pattern, uint8_t CLK, uint8_t MOSI, uint8_t MISO, uint8_t SS) {
+  _SPI_CLK = CLK;
+  _SPI_MOSI = MOSI;
+  _SPI_MISO = MISO;
+  _SPI_SS = SS;
+  begin(row_pattern);
+}
+
+void PxMATRIX::setColorDepth(uint8_t color_depth) {
+  if (color_depth < PxMATRIX_MIN_COLOR_DEPTH) {
+    color_depth = PxMATRIX_MIN_COLOR_DEPTH;
+  }
+  if (color_depth > PxMATRIX_MAX_COLOR_DEPTH) {
+    color_depth = PxMATRIX_MAX_COLOR_DEPTH;
+  }
+  _color_depth = color_depth;
+  _color_half_step = static_cast<int>((256 / _color_depth) / 2);
+}
+
+void PxMATRIX::setSpiFrequency(uint32_t spi_frequency) {
+#if defined(ESP32) || defined(ESP8266)
+  if (spi_frequency < PxMATRIX_MIN_SPI_FREQUENCY) {
+    spi_frequency = PxMATRIX_MIN_SPI_FREQUENCY;
+  }
+  if (spi_frequency > PxMATRIX_MAX_SPI_FREQUENCY) {
+    spi_frequency = PxMATRIX_MAX_SPI_FREQUENCY;
+  }
+  SPI.setFrequency(spi_frequency);
+#endif
+}
+
+uint32_t PxMATRIX::getPixelCount() { return _width * _height; }
+
+void PxMATRIX::spi_init() {
+#ifdef ESP32
+  SPI.begin(_SPI_CLK, _SPI_MISO, _SPI_MOSI, _SPI_SS);
+#else
+  SPI.begin();
+#endif
+
+#if defined(ESP32) || defined(ESP8266)
+  SPI.setFrequency(PxMATRIX_MIN_SPI_FREQUENCY);
+#endif
+
+  SPI.setDataMode(SPI_MODE0);
+  SPI.setBitOrder(MSBFIRST);
+}
+
+void PxMATRIX::begin(uint8_t row_pattern) {
+  _row_pattern = row_pattern;
+  if (_row_pattern == 4) _scan_pattern = ZIGZAG;
+  _mux_pattern = BINARY;
+  _color_order = RRGGBB;
+
+  _pattern_color_bytes = (_height / _row_pattern) * (_width / 8);
+  _row_sets_per_buffer = _rows_per_buffer / _row_pattern;
+  _send_buffer_size = _pattern_color_bytes * 3;
+
+  spi_init();
+
+  pinMode(_OE_PIN, OUTPUT);
+  pinMode(_LATCH_PIN, OUTPUT);
+  pinMode(_A_PIN, OUTPUT);
+  pinMode(_B_PIN, OUTPUT);
+  digitalWrite(_A_PIN, LOW);
+  digitalWrite(_B_PIN, LOW);
+  digitalWrite(_OE_PIN, HIGH);
+
+  if (_row_pattern >= 8) {
+    pinMode(_C_PIN, OUTPUT);
+    digitalWrite(_C_PIN, LOW);
+  }
+  if (_row_pattern >= 16) {
+    pinMode(_D_PIN, OUTPUT);
+    digitalWrite(_D_PIN, LOW);
+  }
+  if (_row_pattern >= 32) {
+    pinMode(_E_PIN, OUTPUT);
+    digitalWrite(_E_PIN, LOW);
+  }
```

Comment:
could be tied together with "else if" to explicitly describe that they're alternative code branches, though i guess the compiler is smart enough to optimize it anyway.